### PR TITLE
Add an instagram embed handler

### LINKED ISSUE #7: Convert Instagram oEmbed and Shortcode to <amp-instagram>
See https://github.com/ampproject/amphtml/blob/master/extensions/amp-instagram/amp-instagram.md

### PR DESCRIPTION
**Pull request by j3j5**
This should close #7

### CODE REVIEW

**Comment by j3j5** [comment]
Before going for #81...is #85 going to be part of the v1 milestone or is this something that's going to be left for later?

**Comment by j3j5** [comment]
The branch is still missing the unit tests and I have to leave now, but I'd thought I'll submit it in case there's any feedback. I'll be AFK till Monday so if somebody wants to make them in the mean time...otherwise, I'll probably push them on Monday.

File: class-amp-instagram-embed.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+require_once( dirname( __FILE__ ) . '/class-amp-embed-handler.php' );
+
+// Much of this class is borrowed from Jetpack embeds
+class AMP_Instagram_Embed_Handler extends AMP_Embed_Handler {
+	const SHORT_URL_HOST = 'instagr.am';
+	const URL_PATTERN = '#http(s?)://(www\.)?instagr(\.am|am\.com)/p/([^/]*)#i';
```
Comment (mjangda) [inline class-amp-instagram-embed.php]: We can update the last selector (`([^/]*)`) to ignore `?` as well using `([^/?]+)` to select the full id.

File: class-amp-instagram-embed.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		$instagram_id = $this->get_instagram_id_from_url( $url );
+
+		return $this->render( array(
+			'url' => $url,
+			'instagram_id' => $instagram_id,
+		) );
+	}
+
+	public function oembed( $matches, $attr, $url, $rawattr ) {
+		return $this->shortcode( array( $url ) );
```
Comment (mjangda) [inline class-amp-instagram-embed.php]: With the updated regex mentioned above, we can bypass the `shortcode` method and just call `render` directly.

File: class-amp-instagram-embed.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			'amp-instagram',
+			wp_parse_args( array(
+				'data-shortcode' => $args['instagram_id'],
+				'layout' => 'responsive',
+			), $this->args )
+		);
+	}
+
+	private function get_instagram_id_from_url( $url ) {
+		$parsed_url = parse_url( $url );
```
Comment (mjangda) [inline class-amp-instagram-embed.php]: We can just pass `PHP_URL_PATH` as the second param and just get back the path given that's the only value we need.

File: class-amp-instagram-embed.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		return array( self::$script_slug => self::$script_src );
+	}
+
+	public function shortcode( $attr ) {
+		$url = false;
+
+		$instagram_id = false;
+		if ( isset( $attr[0] ) ) {
+			$url = ltrim( $attr[0] , '=' );
+		} elseif ( function_exists ( 'shortcode_new_to_old_params' ) ) {
```
Comment (mjangda) [inline class-amp-instagram-embed.php]: The `elseif` isn't needed for this shortcode.

File: class-amp-instagram-embed.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		}
+
+		return array( self::$script_slug => self::$script_src );
+	}
+
+	public function shortcode( $attr ) {
+		$url = false;
+
+		$instagram_id = false;
+		if ( isset( $attr[0] ) ) {
```
Comment (mjangda) [inline class-amp-instagram-embed.php]: I believe the Instagram shortcode expects <a href="https://en.support.wordpress.com/instagram/instagram-images/">an `url` attribute</a>.

**Comment by mjangda** [comment]
Thanks for this @j3j5; looking pretty good! Left a few notes. And unit tests would be great! I've been meaning to write some for the other embeds, but keep putting it off :)

> is #85 going to be part of the v1 milestone or is this something that's going to be left for later?

Yep, will be tackled soon. There are some changes we need to make to all the embed classes (biggest one being moving the shortcode/embed registration out of the constructor and into a method instead).

**Comment by mjangda** [comment]
Thanks again @j3j5. Made a few small changes and merged!

**Pull request merged**